### ISSUE TITLE
Adding support for Wemo Dimmer Switch

### ISSUE THREAD

**Issue by SaunterinRoo** [pywemo/pywemo#77]
I found this project.. Would it be possible to port this into pywemo?

https://github.com/eclipse/smarthome/issues/3632

Thanks! 
Keenan

**Comment by SaunterinRoo** [comment]
What my ultimate goal is, is to see this dimmer switch work with home assistant.

**Comment by sirago** [comment]
I support this issue.

**Comment by stew2000** [comment]
Very impressed with the new wemo dimmers and looking forward to this integration as well!

**Comment by pauldogg** [comment]
I too would love to see the wemo dimmer work with home assistant. I bought 4, let me know if can help with testing...
Thanks

**Comment by pavoni** [comment]
The wemo api is undocumented - so it will need a developer who has a dimmer to add and test it.

With what we already have (and the link above) it shouldn't be too hard - as long as you have the device!

Happy to help.

**Comment by SaunterinRoo** [comment]
I'd also like to add I have the device and would be more than happy to do testing.

-Keenan

**Comment by angel12** [comment]
I have currently gotten basic functionality, as in, pywemo now discovers my dimmer. I will try to work some tonight on adding controls, if I can figure it out!

**Comment by angel12** [comment]
I didnt get any time to work on it in depth tonight, but starting poking at the commands. Is anyone familiar with how I can see what commands are available? I guess I just dont know what I am looking for as far as setting the actual config in the dimmer.py file i put in ouimeaux_device, such as the commands to use / send

**Comment by pavoni** [comment]
Sounds like a great start! When the device is loaded - the code in `api` sets up the functions the device says it supports. So if you load the device in a REPL session - you can use the python inspection functions to poke around.

Take a look at the python functions `dir`, `var` and `keys`. You might even be lucky and get something from `help`

Any wemo device should support `basicevent` but you'll probably find others.

You can then try calling the in the REPL and see what you find.

Good luck!

**Comment by angel12** [comment]
Well, seeing as how I am completely and totally new at python, I guess I need to take a little bit more time in learning it ;) 

All I did to get it discovered was looked at what changes were made to pywemo for the coffee maker, and imitate those changes. If anyone else is more versed in python, feel free to look at my branch of this project, as I have uploaded my changes to it. 

I'll try to get up to speed on python today and potentially spend more time on it this weekend!

**Comment by pavoni** [comment]
It's easier than I've made it sound. 

Your new object will likely have all the functions set up by the discovery process (which is why this is hard to do without owning the device). 

`dir(object)` will list the object's properties and methods - and so give you a good idea about what to call. Not in the same place as my wemo hardware - but here is an example with a string.

```
❯ python3
Python 3.6.3 (default, Oct  4 2017, 06:09:15)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> dir("string")
['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__',
 '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__',
 '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', 
'__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', 
'__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 
'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isdecimal', 'isdigit', 'isidentifier', 'islower',
 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 
'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith',
 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']
>>>
```

The names without __ are methods you can call.

**Comment by angel12** [comment]
so im my case, print(device) when pointed directly to my dimmer, the name is "Living Room". I would replace object in that command with "Living Room" correct? 

```
>>> dir("Living Room")
['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']

```
I poked around some other peoples code that have a basic command set for the dimmer switch working (openhab2 has a pull request that they are working on that has on/off/dimming working for the most part; https://github.com/eclipse/smarthome/pull/3736)

It looks to me that the states that are changed are brightness with values between 0 and 100, and binary state, for on / off. I was hoping to just be able to copy some of their work, even though it's java XD I'm not really a programmer, in the sense that I have no formal training, I just have a decent eye to be able to see what is going on in code haha

**Comment by pavoni** [comment]
try `dir(device).`

From that PR looks like the data is in `basicevent` - so suggest you start with something like the `Switch` logic.

**Comment by angel12** [comment]
```
>>> dir(device)
['WiFiSetup', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_config', '_reconnect_with_device_by_discovery', '_reconnect_with_device_by_probing', '_state', 'basic_state_params', 'basicevent', 'deviceinfo', 'explain', 'firmwareupdate', 'get_service', 'get_state', 'host', 'list_services', 'mac', 'manufacture', 'metainfo', 'model', 'model_name', 'name', 'off', 'on', 'parse_basic_state', 'reconnect_with_device', 'remoteaccess', 'retrying', 'rules', 'serialnumber', 'services', 'set_state', 'smartsetup', 'subscription_update', 'timesync', 'toggle', 'update_binary_state']

```
Thanks for your help ;) I knew that didn't look right

**Comment by angel12** [comment]
So I copied the light switch device file and renamed it to dimmer in my fork, so theoretically I should be able to at least turn the light on / off right? How would I do that from the REPL?

edit: never mind, got it! and it worked!

I am working on getting the dimming functionality fixed up now, and looks like I can get it with preset percentages, now working on getting increments. The big thing I might have problems with is integrating the incremental updates into home-assistant, as that is my end goal. But this is some big progress. I will stop posting here for now, if anyone wants to take a look at my code so far, I have updated my fork on GitHub with the new device configuration with a service of set_dimmer()
